File: cses/palindrom_reorder.cpp
// 0.37s of python over 0.06s of cpp

#include <bits/stdc++.h>

typedef long long ll;

int main() {
   std::string s;
   std::cin >> s;
   std::map<char, ll> map;
   
   for (auto& c: s)
      map[c]++;

   std::string pal = "", rest = "";
   int target = 0;
   char key = ' ';
   ll val = 0;
   for (auto& i: map){
      key = i.first;
      val = i.second;
      if (val % 2 != 0 || val == 1) target++;
      if (target > 1){
         printf("NO SOLUTION\n");
         return 0;
      } 
      if (val % 2 == 0){
         val/=2;
         while(val--)
            pal+=key;
      } else {
         while(val--)
            rest+=key;
      }
   }
   std::string rev_pal = pal;
   std::reverse(rev_pal.begin(), rev_pal.end());
   std::cout << pal << rest << rev_pal << '\n';

}

File: cses/palindrom_reorder.py
n = input()
m = {x:0 for x in n}
for c in n:
    m[c]+=1
s = ''
l = 0
rest = ''
target = 0
for key in m:
    i = m[key]
    if i == 1 or i % 2 != 0:
        target += 1
    if target > 1:
        print("NO SOLUTION")
        quit(0)
    if i % 2 == 0:
        i//=2
        while i:
            i-=1
            s+=key
    else:
        while i:
            i-=1
            rest += key
print(s+rest+s[::-1])
        

